### PR TITLE
Swsh regrid

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Adds a handful of SWSH utilities that I'll need to make future CCE parts easier.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/NumericalAlgorithms/Spectral/SwshTransform.hpp [hunk 1]
```diff
@@ -782,5 +782,69 @@ using make_transform_list_from_derivative_tags =
         tmpl::transform<DerivativeTagList,
                         tmpl::bind<db::remove_tag_prefix, tmpl::_1>>,
         decltype(std::make_integer_sequence<int, 5>{})>::type;
+
+/// \ingroup SwshGroup
+/// \brief Convert spin-weighted spherical harmonic data to a new set of
+/// collocation points (either downsampling or upsampling)
+template <int Spin>
+void interpolate_to_collocation(
```
Comment (moxcodes) [inline src/NumericalAlgorithms/Spectral/SwshTransform.hpp:790]: This should be moved to the cpp

File: src/NumericalAlgorithms/Spectral/SwshTransform.hpp [hunk 2]
```diff
@@ -782,5 +782,69 @@ using make_transform_list_from_derivative_tags =
...
+      number_of_radial_points};
+
+  const auto& target_coefficient_metadata =
+      cached_coefficients_metadata(target_l_max);
+  for(size_t i = 0; i < number_of_radial_points; ++i) {
+    auto source_coefficient_metadata_iterator =
+        cached_coefficients_metadata(source_l_max).begin();
+    for (const auto& coefficient : target_coefficient_metadata) {
+      // note: this assumes the libsharp coefficient ordering for optimizations
+      while (source_coefficient_metadata_iterator !=
```
Comment (kidder) [inline src/NumericalAlgorithms/Spectral/SwshTransform.hpp]: can you add a comment stating what this sets `source_coefficient_metadata_iterator` to so we don't have to figure it out...

File: src/NumericalAlgorithms/Spectral/SwshTransform.hpp [hunk 3]
```diff
@@ -782,5 +782,69 @@ using make_transform_list_from_derivative_tags =
...
+/// \ingroup SwshGroup
+/// \brief Convert spin-weighted spherical harmonic data to a new set of
+/// collocation points (either downsampling or upsampling)
+template <int Spin>
+void interpolate_to_collocation(
+    const gsl::not_null<SpinWeighted<ComplexDataVector, Spin>*> target,
+    const SpinWeighted<ComplexDataVector, Spin>& source,
+    const size_t target_l_max, const size_t source_l_max,
+    const size_t number_of_radial_points) noexcept {
+  auto source_modes =
```
Comment (kidder) [inline src/NumericalAlgorithms/Spectral/SwshTransform.hpp]: should be const

File: src/NumericalAlgorithms/Spectral/SwshFiltering.cpp [hunk 4]
```diff
@@ -34,40 +34,43 @@ void filter_swsh_volume_quantity(
...
-                 *to_filter);
-  const auto& coefficients_metadata = cached_coefficients_metadata(l_max);
-  for (size_t i = 0; i < number_of_radial_grid_points; ++i) {
-    for (const auto& mode : coefficients_metadata) {
-      if (mode.l > limit_l) {
-        transform_buffer->data()[mode.transform_of_real_part_offset +
-                                 i * coefficients_metadata.size()] = 0.0;
-        transform_buffer->data()[mode.transform_of_imag_part_offset +
-                                 i * coefficients_metadata.size()] = 0.0;
+  if (exponential_alpha != 0.0) {
```
Comment (kidder) [inline src/NumericalAlgorithms/Spectral/SwshFiltering.cpp]: use LIKELY

File: src/NumericalAlgorithms/Spectral/SwshFiltering.cpp [hunk 5]
```diff
@@ -34,40 +34,43 @@ void filter_swsh_volume_quantity(
...
+                                          Spectral::Basis::Legendre,
+                                          Spectral::Quadrature::GaussLobatto},
+                                  exponential_alpha, exponential_half_power)),
+                   to_filter->data(),
+                   Index<3>{number_of_swsh_phi_collocation_points(l_max),
+                            number_of_swsh_theta_collocation_points(l_max),
+                            number_of_radial_grid_points});
+    to_filter->data() = *buffer;
+  }
+  if (limit_l < l_max) {
```
Comment (kidder) [inline src/NumericalAlgorithms/Spectral/SwshFiltering.cpp]: and if applicable here LIKELY or UNLIKELY

File: src/NumericalAlgorithms/Spectral/SwshFiltering.cpp [hunk 6]
```diff
@@ -92,17 +95,19 @@ void filter_swsh_boundary_quantity(
...
         transform_buffer) noexcept {
-  transform_buffer->destructive_resize(
-      size_of_libsharp_coefficient_vector(l_max));
-  swsh_transform(l_max, 1, transform_buffer, *to_filter);
-  const auto& coefficients_metadata = cached_coefficients_metadata(l_max);
-  for (const auto& mode : coefficients_metadata) {
-    if (mode.l > limit_l) {
-      transform_buffer->data()[mode.transform_of_real_part_offset] = 0.0;
-      transform_buffer->data()[mode.transform_of_imag_part_offset] = 0.0;
+  if (limit_l < l_max) {
```
Comment (kidder) [inline src/NumericalAlgorithms/Spectral/SwshFiltering.cpp]: again if applicable LIKELY/UNLIKELY

File: src/NumericalAlgorithms/Spectral/SwshCoefficients.cpp [hunk 7]
```diff
@@ -219,6 +251,15 @@ SpinWeighted<ComplexModalVector, Spin> libsharp_to_goldberg_modes(
   template SpinWeighted<ComplexModalVector, GET_SPIN(data)>                   \
   libsharp_to_goldberg_modes(                                                 \
       const SpinWeighted<ComplexModalVector, GET_SPIN(data)>& libsharp_modes, \
+      const size_t l_max) noexcept;                                           \
+  template void goldberg_to_libsharp_modes(                                   \
```
Comment (kidder) [inline src/NumericalAlgorithms/Spectral/SwshCoefficients.cpp]: you only need to instantiate the one that calls the other

**Review by kidder — CHANGES_REQUESTED** [review]
reviewed with @nilsdeppe you can squash

**Comment by moxcodes** [comment]
fixups squashed in.
Cheers!

**Review by nilsdeppe — APPROVED** [review]
image

**Review by kidder — APPROVED** [review]

**Pull request merged**